### PR TITLE
bpo-41317: Remove reader on cancellation in asyncio.loop.sock_accept()

### PR DESCRIPTION
**Pull request by agronholm**
https://bugs.python.org/issue41317

### CODE REVIEW

**Comment by 1st1** [comment]
Looks good to me. @fantix could you please also look at this PR?

File: Lib/asyncio/selector_events.py [hunk 1]
```diff
@@ -555,20 +555,19 @@ async def sock_accept(self, sock):
...
-            self.remove_reader(fd)
-        if fut.done():
-            return
         try:
             conn, address = sock.accept()
             conn.setblocking(False)
         except (BlockingIOError, InterruptedError):
-            self.add_reader(fd, self._sock_accept, fut, True, sock)
+            self._ensure_fd_no_transport(fd)
+            handle = self.add_reader(fd, self._sock_accept, fut, sock)
```
Comment (fantix) [inline Lib/asyncio/selector_events.py]: should be `self._add_reader`

**Review by fantix — CHANGES_REQUESTED** [review]
The rest looks good to me!

File: Lib/asyncio/selector_events.py (hunk 1 above)
Comment (agronholm) [reply Lib/asyncio/selector_events.py]: Fixed.

**Review by fantix — APPROVED** [review]
Thanks!

File: Lib/asyncio/selector_events.py (hunk 1 above)
Comment (1st1) [reply Lib/asyncio/selector_events.py]: I'm curious why this wasn't caught by tests...

Comment (agronholm) [reply Lib/asyncio/selector_events.py]: What exactly? The unprefixed `add_reader()` just calls `self._ensure_fd_no_transport(fd)` before calling `self._add_reader()`, so I could have saved one line by using that and deleting the explicit call to that.

Comment (1st1) [reply Lib/asyncio/selector_events.py]: There's a fundamental difference between them. One returns None, the other returns something else that you appear to use in your code.

Comment (agronholm) [reply Lib/asyncio/selector_events.py]: Right, I see the problem now. Oddly, if I comment out the line that adds the callback, the test **does** trigger the failure.

Comment (agronholm) [reply Lib/asyncio/selector_events.py]: I just tried changing it back to unprefixed `add_reader()` and got a bunch of timeout/cancelled errors in other tests.

Comment (1st1) [reply Lib/asyncio/selector_events.py]: Awesome! The question is now closed then :)

**Comment by 1st1** [comment]
@fantix Thanks for the review!

**Pull request merged**